internal/core/model/adapters.rs

```diff
@@ -413,3 +413,509 @@ fn test_filter_model() {
     assert_eq!(filter.row_data(4).unwrap(), 8);
     assert_eq!(filter.row_count(), 5);
 }
+
+pub trait SortHelper<D> {
+    fn cmp(&mut self, lhs: &D, rhs: &D) -> core::cmp::Ordering;
+}
+
+pub struct AscendingSortHelper;
+
+impl<D> SortHelper<D> for AscendingSortHelper
+where
+    D: core::cmp::Ord,
+{
+    fn cmp(&mut self, lhs: &D, rhs: &D) -> core::cmp::Ordering {
+        lhs.cmp(rhs)
+    }
+}
+
+impl<F, D> SortHelper<D> for F
+where
+    F: FnMut(&D, &D) -> core::cmp::Ordering + 'static,
+{
+    fn cmp(&mut self, lhs: &D, rhs: &D) -> core::cmp::Ordering {
+        (self)(lhs, rhs)
+    }
+}
+
+struct SortModelInner<M, S>
+where
+    M: Model + 'static,
+    S: SortHelper<M::Data>,
+{
+    wrapped_model: M,
+    sort_helper: RefCell<S>,
+    // This vector saves the indices of the elements in sorted order.
+    mapping: RefCell<Vec<usize>>,
+    notify: ModelNotify,
+    sorted_rows_dirty: Cell<bool>,
+}
+
+impl<M, S> SortModelInner<M, S>
+where
+    M: Model + 'static,
+    S: SortHelper<M::Data>,
+{
+    fn build_mapping_vec(&self) {
+        if !self.sorted_rows_dirty.get() {
+            return;
+        }
+
+        let mut mapping = self.mapping.borrow_mut();
+
+        mapping.clear();
+        mapping.extend((0..self.wrapped_model.row_count()).into_iter());
+        mapping.sort_by(|lhs, rhs| {
+            self.sort_helper.borrow_mut().cmp(
+                &self.wrapped_model.row_data(*lhs).unwrap(),
+                &self.wrapped_model.row_data(*rhs).unwrap(),
+            )
+        });
+
+        self.sorted_rows_dirty.set(false);
+    }
+}
+
+impl<M, S> ModelChangeListener for SortModelInner<M, S>
+where
+    M: Model + 'static,
+    S: SortHelper<M::Data>,
+{
+    fn row_changed(&self, row: usize) {
+        if self.sorted_rows_dirty.get() {
+            self.reset();
+            return;
+        }
+
+        let removed_index = self.mapping.borrow().binary_search(&row).unwrap();
+        self.mapping.borrow_mut().remove(removed_index);
```

Comment:
the problem with this algorithm is that in case there are several item which compare the same, this is not going to be a "stable" sort.  I think we would gain in just comparing the item next and after to find out if we need to move the item or not.
Not a blocker for this patch, and could be done in a followup